### PR TITLE
Feature/demo platformer 2d

### PR DESCRIPTION
**Pull request by cogwoz**
Hello!

I'm in the middle of refactoring the platformer demo. You can see a todo list and progress in #57 

I'd like some advice on some things before cracking on again.

1. For the Player and Enemy, should I proceed with a StateMachine as in the Hook demo?
2. I'm having trouble with the moving platforms. To align the Player with them, `sync_to_physics` seems to be necessary but it's not usable with `move_and_slide` so I'm unsure how to do this properly.

Please also offer any other feedback. This is also a learning experience for me as I learn to stick to the guidelines 😄 

Cheers!

### CODE REVIEW

File: platformer-2d-rework/src/Objects/Coin.gd [hunk 1]
```diff
@@ -0,0 +1,11 @@
+extends Area2D
+
+class_name Coin
+
+
+var taken = false
+
+func _on_coin_body_enter(body):
+	if not taken and body is Player:
```
Comment (NathanLovato) [inline platformer-2d-rework/src/Objects/Coin.gd]: You shouldn't need this line, this shows that your physics layer are not right

File: platformer-2d-rework/src/Objects/Coin.gd [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+		($Anim as AnimationPlayer).play("taken")
```
Comment (NathanLovato) [inline platformer-2d-rework/src/Objects/Coin.gd]: Use an onready variable instead to get the node in advance at the top of the script:
```gdscript
onready var animation_player: AnimationPlayer = $AnimationPlayer
```

File: platformer-2d-rework/src/Objects/Coin.gd [hunk 3]
```diff
@@ -0,0 +1,11 @@
+extends Area2D
+
+class_name Coin
+
+
+var taken = false
```
Comment (NathanLovato) [inline platformer-2d-rework/src/Objects/Coin.gd]: Don't add an extra property for that, you can toggle `monitoring` off on the `Area2D`

File: platformer-2d-rework/src/Actors/Enemy.gd [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+func _physics_process(delta):
+	
+	_velocity.x *= -1 if is_on_wall() or not DetectFloorLeft.is_colliding() or not DetectFloorRight.is_colliding() else 1
+
+	_velocity.y = move_and_slide(_velocity, FLOOR_NORMAL).y
+	
+	sprite.scale.x = 1 if _velocity.x > 0 else -1
+	
+	# TODO: Add separate StateMachine?
```
Comment (NathanLovato) [inline platformer-2d-rework/src/Actors/Enemy.gd]: No, that'd be too much for this demo. There's a demo dedicated to state machines iirc

File: platformer-2d-rework/src/Actors/Enemy.gd [hunk 5]
```diff
@@ -0,0 +1,44 @@
+extends Actor
+
+class_name Enemy
+
+const STATE_WALKING = 0
```
Comment (NathanLovato) [inline platformer-2d-rework/src/Actors/Enemy.gd]: Use an enum, and keep names consistent. Either -ing verbs or nouns that describe the current state, or action verbs (e.g. walking/jumping/dead or walk/jump/die)
```gdscript
enum State {WALKING, DEAD}
```

File: platformer-2d-rework/src/Actors/Player.gd [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+class_name Player
+
+const BULLET_VELOCITY := 1000.0
+const SHOOT_TIME_SHOW_WEAPON := 0.2
+
+var _current_animation := ""
+var _shoot_time := 99.0 # time since last shot
+
+onready var sprite = $Sprite
```
Comment (NathanLovato) [inline platformer-2d-rework/src/Actors/Player.gd]: onready vars should be at the top. Also they should use type hints

File: platformer-2d-rework/src/Actors/Player.gd [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+	# Shooting
+	# TODO: Move this to a separate Gun scene?
+	if Input.is_action_just_pressed("shoot"):
+		var bullet = Bullet.instance()
+		bullet.position = ($Sprite/BulletShoot as Position2D).global_position # use node for shoot position
+		bullet.linear_velocity = Vector2(sprite.scale.x * BULLET_VELOCITY, 0)
+		bullet.add_collision_exception_with(self) # don't want player to collide with bullet
+		get_parent().add_child(bullet) # don't want bullet to move with me, so add it as child of parent
+		($SoundShoot as AudioStreamPlayer2D).play()
```
Comment (NathanLovato) [inline platformer-2d-rework/src/Actors/Player.gd]: Use onready vars and avoid casting

File: platformer-2d-rework/src/Actors/Player.gd [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+	var snap: Vector2 = Vector2.DOWN * 60.0 if direction.y == 0.0 else Vector2.ZERO
+	_velocity = move_and_slide_with_snap(
+		_velocity, snap, FLOOR_NORMAL, true, 4,  0.885398, false
+	)
+
+	# Shooting
+	# TODO: Move this to a separate Gun scene?
+	if Input.is_action_just_pressed("shoot"):
+		var bullet = Bullet.instance()
+		bullet.position = ($Sprite/BulletShoot as Position2D).global_position # use node for shoot position
```
Comment (NathanLovato) [inline platformer-2d-rework/src/Actors/Player.gd]: Use onready var and avoid casting

**Review by NathanLovato — DISMISSED** [review]
Quick review on a handful of points. I don't know if you changed some code already but looking at it, some doesn't look like code from the official demos.

I don't think a lost/win overlay is necessary. Having just one level with the playground is enough. The goal of these demos isn't to have full games I think, but to show how a feature works, or how to implement a mechanic, or in this case the core loop of a specific type of game. Doing more will make the demos overlap and use a lot more production + maintenance time.

General notes:

- We get most nodes using the `onready` keyword. The goal is to have all dependencies at the top of the scripts.
- You don't need casting that way
- Avoid shorthands in variable names, e.g. favor `animation` over `anim`

**Comment by NathanLovato** [comment]
Just fixed the moving platform issue. Here's the explanation:
```
The issue was that the floor's velocity (platform here) is only transferred to
the character if move_and_slide_with_snap's stop_on_slope argument is `false`.

stop_on_slope seems to stop the character's movement if it has no horizontal
velocity and the floor's normal is < a threshold angle to the floor_normal
parameter, even if it's horizontal in our case.

To preserve stop_on_slope on slopes, I've introduced a Raycast2D that only
detects platforms.
```

**Comment by NathanLovato** [comment]
Regarding the code style/guidelines, would you like me to do e.g. part of a script, or part of the project to give you a headstart + example? The code has some no-nos to me at this point.

**Comment by cogwoz** [comment]
Thanks for that Nathan. 

There are some parts of the code that are left over from the original demo that I haven't touched yet. I was planning on going over the scripts in more detail after everything was working using the guidelines but an example would be helpful if you have time.

**Comment by NathanLovato** [comment]
There you go, I just rewrote the code for the Player script.
There's more refactoring to do to show good practices: the physics layers and masks are a mess, so the original code relied on things like having the bullet get its parent and making it an exception for collisions.
Anyway, the player is the most complicated bit here. from there, I think we could make some more gameplay improvements, but it'll be part of the polishing phase.
I'd port all the code to fit the guidelines at this point.

Regarding moving platforms, I'd also remove the code to use an animation player too, as that way, you can directly move the platforms in the editor and even use the new animation autokey feature to make the platforms move exactly where you want.

**Comment by cogwoz** [comment]
That's great thanks for that! I'll work on the rest of the scenes today.

**Comment by cogwoz** [comment]
The other scenes should be better now and I've tidied up the stage.

I wanted to use Path2D for the moving platforms as I thought moving the path in the editor would make intuitive sense but the paths stop working when the KinematicBody2D is set to `sync_to_physics` so I used an AnimationPlayer instead. Still sad about that one.

**Comment by NathanLovato** [comment]
I just tweaked some tweaks. It's looking great to me! Do you still want to change anything or is it good to merge?

**Comment by cogwoz** [comment]
Great! If you're happy with everything then it's all good!

**Pull request merged**